File: Three.java
import mpi.*;
import java.io.*;
import java.nio.file.*;
import java.util.*;

public class FileSumMPI {
    public static void main(String[] args) throws Exception {
        MPI.Init(args);

        int rank = MPI.COMM_WORLD.Rank();
        int size = MPI.COMM_WORLD.Size();

        // Массив имен файлов (можно передавать как аргументы)
        String[] filenames = {"file1.txt", "file2.txt", "file3.txt", "file4.txt"};
        int n = filenames.length;

        if (n < size) {
            if (rank == 0) {
                System.err.println("Number of files (" + n + ") is less than number of processes (" + size + ")");
            }
            MPI.Finalize();
            return;
        }

        // Распределение файлов по процессам
        int filesPerProc = n / size;
        int extraFiles = n % size;

        int startFile, endFile;
        if (rank < extraFiles) {
            startFile = rank * (filesPerProc + 1);
            endFile = startFile + filesPerProc + 1;
        } else {
            startFile = extraFiles * (filesPerProc + 1) + (rank - extraFiles) * filesPerProc;
            endFile = startFile + filesPerProc;
        }

        // Каждый процесс вычисляет сумму для своих файлов
        double localSum = 0.0;
        for (int i = startFile; i < endFile; i++) {
            localSum += calculateFileSum(filenames[i]);
            if (rank == 0) {
                System.out.println("Process " + rank + " processed file " + filenames[i]);
            }
        }

        // Сбор результатов на процессе 0
        double[] totalSum = new double[1];
        MPI.COMM_WORLD.Reduce(new double[]{localSum}, 0, totalSum, 0, 1, MPI.DOUBLE, MPI.SUM, 0);

        // Процесс 0 записывает результат в файл
        if (rank == 0) {
            try (PrintWriter out = new PrintWriter("res.txt")) {
                out.println(totalSum[0]);
                System.out.println("Total sum: " + totalSum[0] + " written to res.txt");
            } catch (IOException e) {
                System.err.println("Error writing to res.txt: " + e.getMessage());
            }
        }

        MPI.Finalize();
    }

    private static double calculateFileSum(String filename) {
        try {
            return Files.lines(Paths.get(filename))
                    .flatMap(line -> Arrays.stream(line.split("\\s+")))
                    .filter(s -> !s.isEmpty())
                    .mapToDouble(Double::parseDouble)
                    .sum();
        } catch (IOException | NumberFormatException e) {
            System.err.println("Error processing file " + filename + ": " + e.getMessage());
            return 0.0;
        }
    }
}

File: Three.cpp

#include <iostream>
#include <fstream>
#include <vector>
#include <string>
#include <mpi.h>

using namespace std;

double calculate_file_sum(const string& filename) {
    ifstream file(filename);
    if (!file.is_open()) {
        cerr << "Error opening file: " << filename << endl;
        return 0.0;
    }

    double sum = 0.0;
    double num;
    while (file >> num) {
        sum += num;
    }
    file.close();
    return sum;
}

int main(int argc, char** argv) {
    MPI_Init(&argc, &argv);

    int rank, size;
    MPI_Comm_rank(MPI_COMM_WORLD, &rank);
    MPI_Comm_size(MPI_COMM_WORLD, &size);

    // Массив имен файлов (в реальной программе можно считать из входных данных)
    const char* filenames[] = {"file1.txt", "file2.txt", "file3.txt", "file4.txt"};
    const int n = sizeof(filenames) / sizeof(filenames[0]);

    if (n < size) {
        if (rank == 0) {
            cerr << "Number of files (" << n << ") is less than number of processes (" << size << ")" << endl;
        }
        MPI_Finalize();
        return 1;
    }

    // Распределяем файлы по процессам
    int files_per_proc = n / size;
    int extra_files = n % size;

    int start_file, end_file;
    if (rank < extra_files) {
        start_file = rank * (files_per_proc + 1);
        end_file = start_file + files_per_proc + 1;
    } else {
        start_file = extra_files * (files_per_proc + 1) + (rank - extra_files) * files_per_proc;
        end_file = start_file + files_per_proc;
    }

    // Каждый процесс вычисляет сумму для своих файлов
    double local_sum = 0.0;
    for (int i = start_file; i < end_file; i++) {
        local_sum += calculate_file_sum(filenames[i]);
        if (rank == 0) {
            cout << "Process " << rank << " processed file " << filenames[i] << endl;
        }
    }

    // Собираем результаты на процессе 0
    double total_sum = 0.0;
    MPI_Reduce(&local_sum, &total_sum, 1, MPI_DOUBLE, MPI_SUM, 0, MPI_COMM_WORLD);

    // Процесс 0 записывает результат в файл
    if (rank == 0) {
        ofstream outfile("res.txt");
        if (outfile.is_open()) {
            outfile << total_sum;
            outfile.close();
            cout << "Total sum: " << total_sum << " written to res.txt" << endl;
        } else {
            cerr << "Error opening output file res.txt" << endl;
        }
    }

    MPI_Finalize();
    return 0;
}
